### PR TITLE
Add Expansion UX, Rules, Setting, Learning Quest

### PR DESCRIPTION
**Pull request by toddmedema**
Closes https://github.com/ExpeditionRPG/expedition-app/issues/386

On top of adding all of the new expansion UI things, this also fixes some small annoying UI bugs, primarily that picker titles had their line height set really large to vertically center the text, but that broke when the text became two lines. Looks much better now! Also improves the display of dialogs (adding the papyrus background + better button styling), and the combat damage page (highlights damage, and doesn't show the "knocked out?" help text unless alive adventurers < players).

screen shot 2017-09-19 at 2 42 21 pm
screen shot 2017-09-19 at 2 42 45 pm
screen shot 2017-09-19 at 2 23 14 pm
screen shot 2017-09-19 at 2 15 39 pm
screen shot 2017-09-19 at 2 14 27 pm
screen shot 2017-09-19 at 2 41 43 pm
screen shot 2017-09-19 at 2 05 09 pm
screen shot 2017-09-19 at 2 04 54 pm

### CODE REVIEW

**Comment by toddmedema** [comment]
Improved prompt subtext: 
screen shot 2017-09-21 at 8 19 30 pm

File: app/Globals.test.tsx [hunk 1]
```diff
@@ -29,4 +29,17 @@ describe('Globals', () => {
...
   });
+
+  describe('getStorageKey', () => {
+    it('Boolean');
+    it('JSON');
+    it('Number');
+    it('String');
+  });
+  describe('setStorageKey', () => {
+    it('Converts booleans to strings');
```
Comment (smartin015) [inline app/Globals.test.tsx]: No caps on start of it() string

File: app/Globals.tsx [hunk 2]
```diff
@@ -91,9 +91,15 @@ export function setNavigator(navigator: any) {
   refs.navigator = navigator;
 }
 
+// accepts boolean, numbers, strings and JSON containing only those types
```
Comment (smartin015) [inline app/Globals.tsx]: Comment doesn't match the function implementation

File: app/Globals.tsx [hunk 3]
```diff
@@ -91,9 +91,15 @@ export function setNavigator(navigator: any) {
...
 export function setStorageKey(key: string, value: any) {
```
Comment (smartin015) [inline app/Globals.tsx]: `setStorageItem` maybe? This isn't just setting a key.

File: app/Globals.tsx [hunk 4]
```diff
@@ -91,9 +91,15 @@ export function setNavigator(navigator: any) {
...
   try {
-    getWindow().localStorage.setItem(key, value.toString());
+    if (typeof value === 'object') {
+      value = JSON.stringify(value);
+    } else {
+      value = value.toString();
+    }
+    getWindow().localStorage.setItem(key, value);
   } catch (e) {
     console.log('Error setting storage key', key, 'to', value, e);
```
Comment (smartin015) [inline app/Globals.tsx]: `console.error`

File: app/Globals.tsx [hunk 5]
```diff
@@ -129,6 +135,11 @@ export function getStorageKeyBoolean(key: string, fallback: boolean): boolean {
   return (val !== null) ? (val.toLowerCase() === 'true') : fallback;
 }
 
+export function getStorageKeyJson(key: string, fallback?: object): object {
```
Comment (smartin015) [inline app/Globals.tsx]: Same here... the returned values aren't keys

File: app/Main.tsx [hunk 6]
```diff
@@ -195,6 +196,14 @@ export function init() {
   setupGoogleAnalytics();
   getStore().dispatch(fetchAnnouncements());
 
+  const contentSets = (getStore().getState().settings || {}).contentSets;
+  for (const set in contentSets) {
```
Comment (smartin015) [inline app/Main.tsx]: So if settings fails to load, the for loop and modal are skipped? Is this the behavior we want?

File: app/cardtemplates/combat/Combat.tsx [hunk 7]
```diff
@@ -211,9 +211,11 @@ function renderSurge(props: CombatProps): JSX.Element {
 
 function renderResolve(props: CombatProps): JSX.Element {
   let helpText: JSX.Element = (<p>Resolve all played abilities.</p>);
+  const theHorror = (props.settings.contentSets.horror === true);
   if (props.settings.showHelp) {
     helpText = (
       <span>
+        {theHorror && <Callout icon="horror_white"><strong>The Horror:</strong> Any adventurers at Min persona must resolve their persona effect and then return to Base before resolving abilities. Adventurers at Max persona may choose to resolve and reset now or in a later round.</Callout>}
```
Comment (smartin015) [inline app/cardtemplates/combat/Combat.tsx]: `to Base persona`

File: app/cardtemplates/combat/Combat.tsx [hunk 8]
```diff
@@ -296,6 +300,7 @@ function renderPlayerTier(props: CombatProps): JSX.Element {
 
 function renderVictory(props: CombatProps): JSX.Element {
   const contents: JSX.Element[] = [];
+  const theHorror = (props.settings.contentSets.horror === true);
```
Comment (smartin015) [inline app/cardtemplates/combat/Combat.tsx]: I'm seeing this repeated quite a bit. Is there a way we can generalize this? Export a `horrorExpansion()` boolean function somewhere?

File: app/cardtemplates/combat/State.test.tsx [hunk 9]
```diff
@@ -2,19 +2,23 @@
 describe('Combat State', () => {
   describe('combatScope', () => {
     describe('randomEnemy', () => {
-      it('gets a random enemy');
+      it('gets a random enemy name');
```
Comment (smartin015) [inline app/cardtemplates/combat/State.test.tsx]: No implementation? :(

File: app/Globals.tsx (hunk 2 above)
Comment (toddmedema) [reply app/Globals.tsx]: Hmm? That's what value can be. Updated the comment to "Value can be boolean, number, string or stringifiable JSON" to increase clarity

File: app/Globals.tsx (hunk 3 above)
Comment (toddmedema) [reply app/Globals.tsx]: Updated to setStorageKeyValue since that seems to be the clearest description of what it's doing :)

File: app/Globals.tsx (hunk 5 above)
Comment (toddmedema) [reply app/Globals.tsx]: Reduced to `getStorageJson`, etc

File: app/Main.tsx (hunk 6 above)
Comment (toddmedema) [reply app/Main.tsx]: Good point; updated to show the modal if settings fail to load.

File: app/cardtemplates/combat/Combat.tsx (hunk 8 above)
Comment (toddmedema) [reply app/cardtemplates/combat/Combat.tsx]: Not sure that's worth it at this point since we'd be replacing a boolean check with a function call, and then have to worry about managing that function. Perhaps we can design a more generalizable solution in 6 months once we have multiple expansions and keeping track of them gets more complex :)

File: app/cardtemplates/combat/State.test.tsx (hunk 9 above)
Comment (toddmedema) [reply app/cardtemplates/combat/State.test.tsx]: Ask and you shall receive :)

**Comment by toddmedema** [comment]
@smartin015 all comments addressed!

File: app/Globals.tsx (hunk 2 above)
Comment (smartin015) [reply app/Globals.tsx]: As implemented, it'll store anything that's an object type or has a toString function (i.e. it doesn't explicitly reject Array, or Set, etc). JSON is a string wire format, not a javascript object.

If you want to restrict to those types, update value param to be `number|boolean|Object` instead of `any`.

File: app/Main.tsx [hunk 10]
```diff
@@ -196,12 +196,17 @@ export function init() {
...
+  if (settings) {
+    const contentSets = (settings || {}).contentSets;
+    for (const set in contentSets) {
+      if (contentSets[set] === null) {
+        getStore().dispatch(setDialog('EXPANSION_SELECT'));
+        break;
+      }
     }
+  } else {
+    getStore().dispatch(setDialog('EXPANSION_SELECT'));
```
Comment (smartin015) [inline app/Main.tsx]: Can you add a test that asserts this dialog shows if settings is null and/or contentSets is weird?
